src/types.rs

```diff
@@ -299,6 +299,35 @@ impl Type {
             }
         }
     }
+
+    /// Could make this public, but easier to panic if the TypeArg's don't match the TypeParams
+    /// (should be checked when the args are first given to the type scheme).
+    pub(crate) fn substitute(&self, args: &[TypeArg]) -> Self {
```

Comment:
I tried this, but found it barely helped (if at all), and has to be less efficient - clone first clones *everything*, then builds some new bits where the old bits have to change; substitute-takes-ref only clones the bits that are common to the original, and then builds the same "new bits". See dd00c63220d85449773141a0320ff1a1bc7f4e58 if you want to take a look (but I've then reverted).

Moreover, the common case is in a type scheme's `compute_signature`, in which case we can't consume the type scheme, so we'd have to clone. (Possibly at some point we may use this for substituting across a whole Hugr, not just the types, and *possibly* the code to do that could be written as a destructive mutator rather than a copy-and-create, that might fit well with a Tierkreis/rust-style implementation, so there might be some benefit there in being able to consume the input Hugr. Seems very speculative at this point though.)